File: 110-balanced-binary-tree/110-balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

const checkNode = function (node) {
    if (!node) return [0, true];
    
    const [lh, lbs] = checkNode(node.left);
    if (!lbs) return [0, false];
 
    const [rh, rbs] = checkNode(node.right);
    if (!rbs) return [0, false];
    
    
    const balanced = Math.abs(lh - rh) <= 1;
    
    const maxHeight = Math.max(lh, rh);
    
    return [maxHeight + 1, balanced];
    
}

const isBalanced = function(root) {
    return checkNode(root)[1];
};

File: 110-balanced-binary-tree/110-balanced-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    return is_balanced_helper(root)[0]
end

def is_balanced_helper(node)
    return [true, 0] unless node
    
    right_balanced, right_max_depth = is_balanced_helper(node.right)
    left_balanced, left_max_depth = is_balanced_helper(node.left)
    
    balanced = left_balanced && right_balanced 
    return [false, 0] unless balanced
    
    balanced = (left_max_depth - right_max_depth).abs <= 1
    return [false, 0] unless balanced
    
    max_depth = [left_max_depth, right_max_depth].max + 1
    
    return [true, max_depth]
end
    